build/report/report.go

```diff
@@ -0,0 +1,220 @@
+  </body>
+</html>  
+`
+)
+
+type report struct {
+	WindowStart  string
+	WindowEnd    string
+	Flakes       []flake
+	Builds       int
+	FlakePercent int
+}
+
+type flake struct {
+	ID         string
+	CreateTime string
+}
+
+type redirect struct {
+	Date string
+}
+
+func main() {
+	ctx := context.Background()
+	reportTmpl := newReportTemplate()
+	redirTmpl := newRedirectTemplate()
+
+	windowEnd := time.Now().UTC()
+	windowStart := windowEnd.Add(-window)
+	date := windowEnd.Format("2006-01-02")
+
+	err := os.MkdirAll(outPath, 0o755)
+	if err != nil {
+		log.Fatalf("failed to create output path %v: %v", outPath, err)
+	}
+
+	datePath := fmt.Sprintf("%s/%s.html", outPath, date)
+	reportFile, err := os.OpenFile(datePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		log.Fatalf("failed to open report %v: %v", datePath, err)
+	}
+
+	redirPath := fmt.Sprintf("%s/index.html", outPath)
+	redirFile, err := os.OpenFile(redirPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		log.Fatalf("failed to open redirect %v: %v", redirPath, err)
+	}
+
+	c, err := cloudbuild.NewClient(ctx)
+	if err != nil {
+		log.Fatalf("failed to initialize cloudbuild client: %v", err)
+	}
+
+	success := make(map[string]bool)     // build SHA -> bool
+	failure := make(map[string][]string) // build SHA -> slice of build IDs that failed
+	idTime := make(map[string]time.Time) // build ID -> create time
+
+	// See https://pkg.go.dev/cloud.google.com/go/cloudbuild/apiv1/v2/cloudbuildpb#ListBuildsRequest.
+	req := &cloudbuildpb.ListBuildsRequest{
+		ProjectId: "agones-images",
+		// TODO(zmerlynn): No idea why this is failing.
+		// Filter:    `build_trigger_id = "da003bb8-e9bb-4983-a556-e77fb92f17ca"`,
+	}
+	it := c.ListBuilds(ctx, req)
+	for {
+		resp, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			log.Fatalf("error listing builds: %v", err)
+			break
+		}
+		createTime := resp.CreateTime.AsTime()
+		if createTime.Before(windowStart) {
+			break
+		}
+		// We only care about Agones builds.
+		if resp.BuildTriggerId != wantBuildTriggerID {
+			continue
+		}
+		// Ignore if it's still running.
+		if resp.FinishTime == nil {
+			continue
+		}
+
+		id := resp.Id
+		sha := resp.Substitutions["COMMIT_SHA"]
+		status := resp.Status
+		idTime[id] = createTime
+		log.Printf("id = %v, sha = %v, status = %v", id, sha, status)
+
+		// Record clear cut success/failure, not timeout, cancelled, etc.
+		switch status {
+		case cloudbuildpb.Build_SUCCESS:
+			success[sha] = true
+		case cloudbuildpb.Build_FAILURE:
+			failure[sha] = append(failure[sha], id)
+		default:
+			continue
+		}
+	}
+
+	buildCount := 0
+	flakeCount := 0
+	var flakes []flake
+	for sha := range success {
+		buildCount++
+		if ids, ok := failure[sha]; ok {
+			flakeCount++
+			for _, id := range ids {
+				flakes = append(flakes, flake{
+					ID:         id,
+					CreateTime: idTime[id].Format(time.RFC3339),
+				})
+			}
+		}
+	}
```

Comment:
```suggestion
	buildCount := len(success)
	for sha := range success {	
		ids, ok := failure[sha]; ok
		if !ok {
			continue
		}
		flakeCount++
		for _, id := range ids {
			flakes = append(flakes, flake{
			    ID:         id,
			    CreateTime: idTime[id].Format(time.RFC3339),
			})
		}
	}
```

A suggestion, I don't know if it would be better, at least not so deep

Reply:
Thanks for the suggestion. I decided to change methodology and just count all the flakes and report the flake ratio, which makes this less complicated anyways.